Airtable migration

Robert-Roth opened the pull request:
As part of our move to PubPub Platform, we'd like to store Twitter collections in Airtable rather than Notion to simplify our workflows and allow for automatic syncing to PubPub Platform. We've modified this repository to reflect that.

- Switched all Notion calls to Airtable calls
- Modified workflow files to call Airtable secrets and modified the create_collections cron to trigger on a repository dispatch event
- Added more detailed logging throughout, as the Twitter API is brittle and it's helpful to see more verbose output
- Added more explicit error handling in some scripts (e.g., clean_old_collections.py). I also think this script could probably be removed — appears to be a legacy tool. It appears that its functionality already exists when tweets are gathered
- Added tests to help evaluate these changes — tweet volume is low after the first week of publication, so it is difficult to test with live data
- Updated README

I think in the future we should try to make this a bit cleaner and have even better error handling, but these changes should allow an uninterrupted migration for now.

keithchev commented on src/fetch_tweets.py:
```diff
@@ -2,76 +2,94 @@
-        collection_id = get_rich_text_value(entry, "ID")
+        collection_id = get_field_value(entry, "ID")
         if collection_id:
-            # Priority represents recency in a decaying fashion. Something that is just created has
-            # a priority of 1. Something that has been created 2 years ago, will have a priority of
-            # close to 0. The priority dictates the frequency of fetching tweets. So, something that
-            # is just created will have a higher chance of fetching tweets than something that is
-            # 2 years old.
```
I would leave this comment as-is, because it looks like it's still relevant and it's helpful. Though it could maybe be moved to the docstring of `calculate_priority`.

keithchev commented on src/fetch_quote_tweets.py:
```diff
@@ -2,64 +2,86 @@
-        collection_id = get_rich_text_value(entry, "ID")
+        collection_id = get_field_value(entry, "ID")
         if collection_id:
-            # Priority represents recency in a decaying fashion. Something that is just created has
-            # a priority of 1. Something that has been created 2 years ago, will have a priority of
-            # close to 0. The priority dictates the frequency of fetching tweets. So, something that
-            # is just created will have a higher chance of fetching tweets than something that is
-            # 2 years old.
```
Same comment as above - I would keep this comment but move it to `calculate_priority` so that it's not duplicated here and in `fetch_tweets.py`.

keithchev commented on tests/test_fetch_tweets.py:
```diff
@@ -0,0 +1,84 @@
+import os
+import sys
+
+# Add the src directory to Python path so I can avoid restructuring this project
+TEST_DIR = os.path.dirname(os.path.abspath(__file__))
+SRC_DIR = os.path.join(os.path.dirname(TEST_DIR), 'src')
+sys.path.insert(0, SRC_DIR)
```
A simpler way of handling this is to add a `pytest.ini` file to the repo with the lines:
```
[pytest]
pythonpath = src
```

keithchev commented on tests/test_fetch_tweets.py:
```diff
@@ -0,0 +1,84 @@
+    airtable.get_database_entries = Mock(return_value=[entry])
+    airtable.update_page = Mock()
+
+    return {'twitter': twitter, 'airtable': airtable, 'entry': entry}
```
Nit: it's cleaner and more conventional to return a tuple:
```suggestion
    return twitter,  airtable, entry
```

keithchev commented on tests/test_fetch_tweets.py:
```diff
@@ -0,0 +1,84 @@
+        'id': '1734567890123456791',
+        'text': 'New tweet! https://research.arcadiascience.com/pub/test-pub'
+    }
+
```
If the `mock_apis` fixture returns a tuple, you can unpack it here like this:
```suggestion
twitter, airtable, entry = mock_apis
```

keithchev commented on tests/test_fetch_quote_tweets.py:
```diff
@@ -0,0 +1,123 @@
+sys.path.insert(0, SRC_DIR)
+
+
+@pytest.fixture
+def mock_apis():
```
To avoid duplicating this fixture in both test files, it can be moved to a file called `tests/conftest.py`. This is the conventional place to put shared fixtures (and it is automatically loaded by pytest).

keithchev commented on tests/test_fetch_quote_tweets.py:
```diff
@@ -0,0 +1,123 @@
+    return {'twitter': twitter, 'airtable': airtable, 'entry': entry}
+
+
+def test_quote_tweet_collection(mock_apis):
```
Nit: for functions that test a specific function, it's good to name the test function exactly after the function it is testing:
```suggestion
def test_get_quote_tweets(mock_apis):
```

keithchev commented on tests/test_fetch_quote_tweets.py:
```diff
@@ -0,0 +1,123 @@
+    assert actual_tweets == expected_tweets
+
+
+def test_quote_tweet_error_handling(mock_apis):
```
Same comment about naming
```suggestion
def test_get_quote_tweets_error_handling(mock_apis):
```

keithchev commented on tests/test_fetch_quote_tweets.py:
```diff
@@ -0,0 +1,123 @@
+    assert actual_tweets == expected_tweets
+
+
+def test_no_quote_tweets_found(mock_apis):
```
```suggestion
def test_get_quote_tweets_no_quote_tweets_found(mock_apis):
```

keithchev commented on tests/test_fetch_tweets.py:
```diff
@@ -0,0 +1,84 @@
+    return {'twitter': twitter, 'airtable': airtable, 'entry': entry}
+
+
+def test_tweet_collection_and_storage(mock_apis):
```
Same comment about naming tests after the function being tested
```suggestion
def test_get_tweets_for_entry(mock_apis):
```

keithchev commented on tests/test_fetch_tweets.py:
```diff
@@ -0,0 +1,84 @@
+    assert actual_tweets == expected_tweets
+
+
+def test_error_handling_and_fallback(mock_apis):
```
```suggestion
def test_get_tweets_for_entry_error_handling(mock_apis):
```

keithchev reviewed (changes requested):
Looks good! The tests are a nice addition. I made a few suggestions and edits, all of them minor.

Robert-Roth commented on README.md:
```diff
@@ -44,20 +50,43 @@ If you have any questions about the implementation please reach out to [mertcele
+# The ID of the base where your Twitter collections table is stored
+AIRTABLE_BASE_ID=
+
+# The ID or name of the table that holds your Twitter collections
```
Airtable allows you to use the table name or ID interchangeably, but it's best practice to use the ID (since it's persistent can't be changed like the name can). Do you think we should adjust this to AIRTABLE_TABLE_ID and recommend that people use the ID, or does this suffice?

Robert-Roth replied:
Actually, thinking about this more we should change everything to AIRTABLE_TABLE_ID. It's more maintainable, doesn't require any URL encoding, and is just all-around better. Going to make those changes.

Robert-Roth reviewed (commented):
@keithchev I went ahead and made the changes you suggested (everything seemed great, thank you!). I believe the only outstanding thing is my reply about table IDs in the README. Let me know if there's anything else I should change.

Robert-Roth commented:
I changed the TABLE_NAME to a TABLE_ID for better maintainability. While I was in there, I also changed everything to use a field ID instead of a name so that we (or anyone using this) can freely change their field names. Sorry to add more work here! We can revert to using field names if you prefer, but I figured this would be better in the long run.

keithchev commented on README.md:
```diff
@@ -62,7 +62,12 @@ If you have any questions about the implementation please reach out to [mertcele
-> **ID**: str, automatically generated by the script
+> **ID**: str, automatically generated by the script. Note that this is distinct from Airtable's internal record ID.
+
+### Edit the `field_constants.py` file
```
It sounds like the field IDs are specific to the airtable base corresponding to `AIRTABLE_BASE_ID` (and maybe also the table?), so I think the field IDs should be environment variables rather than constants, because they're deployment- or environment-specific. It's also a bit confusing to have configuration settings split between the env file and hard-coded constants. You could still use the `Fields` class as-is, it would just need to be instantiated from the environment variables at runtime.

Robert-Roth replied:
I like this idea. They are specific to the base and the table, so it makes sense to have it be in the env. I made changes with this suggestion to keep it as easily configurable as possible.

keithchev commented on README.md:
```diff
@@ -62,7 +62,12 @@ If you have any questions about the implementation please reach out to [mertcele
+
+### Edit the `field_constants.py` file
+The Airtable API can return either the name of the field or the field ID. It is recommended to use the field ID in case
+you want to change the name of the field in Airtable for any reason. As such, use the [web API documentation](https://airtable.com/developers/web/api/introduction) to 
+find the field IDs of the fields above and alter the `field_constants.py` file as needed. 
```
Is it possible to point to a more specific part of the API docs? I couldn't immediately figure out how to retrieve the field IDs for a given table. Is it the "get base schema" endpoint ([here](https://airtable.com/developers/web/api/get-base-schema)?

Robert-Roth replied:
This one is a little tricky. The link right now points to their web API. If you have the necessary permissions, you can view the API docs as generated for your specific base (with all the field IDs, table IDs, base IDs, etc.). The one you linked to will give the same information, but via API requests. So the link I have there is the 'friendlier' version, but it does require some navigation (selecting your base > navigating to the specific table)

keithchev reviewed (changes requested):
Looks good! I agree it could be more maintainable to use field IDs rather than field names, but I think the field IDs would have to live in the env file, not in the source code, since they're specific to the `BASE_ID` in the env file.

keithchev commented:
Another thought: one alternative for the field IDs is to retain the `Fields` class but to set its attributes to be the field names, rather than the IDs. This avoids hard-coding the field names (like `"ID"`) as string literals everywhere, and it also avoids the need to specify all of the field IDs as env variables. I think this is a fine compromise; as long as the README makes it clear that the table corresponding to `TABLE_ID` must have columns with the specific exact names listed in the `Fields` class.

Robert-Roth commented:
I moved field IDs to the env file (I think this makes the most sense) and adjusted the README. Also removed some unused code from airtable.py.
